core/frontend/src/Marker.ts

```diff
@@ -443,3 +457,106 @@ export abstract class MarkerSet<T extends Marker> {
     }
   }
 }
+
+/**
+ * A [[MarkerSet]] sub-class that uses a gready clustering algorithm. Sub-classes should use the [[getAverageLocation]] function
```

Comment:
gready -> greedy

Please explain the algorithm here (I see it below, don't understand why it's called "greedy".) Have you tested the performance of this for large marker sets? It seems to me that it will be substantially slower and I'm not sure when someone should use it.

Reply:
i'd like to understand the problem you're trying to solve (the "unexpected and noticeable" part). Maybe you can make a video showing the difference between the two algorithms? The whole reason for the current algorithm is to avoid all the n2 calculations you've added here - this is obviously all done per-frame during view manipulations.

Reply:
did you try just leave the current clustering algorithm but use the middle of the union of the ranges of the members when you create the cluster Marker?

Reply:
The algorithm comes from here: https://blog.mapbox.com/clustering-millions-of-points-on-a-map-with-supercluster-272046ec5c97

I'm not doing the hierarchical portion of that algorithm because I'm not retaining the results as an index for different zoom levels. I think it's called "greedy" as it clusters everything at the chosen point within the radius. I see that they call it supercluster in their [github repo](https://github.com/mapbox/supercluster).

As shown in the code comments, the algorithm itself is simply:

- Start with any point from the dataset.
  - I start with the first visible marker
- Find all points within a certain radius around that point.
  - I iterate over all the other visible markers that we haven't already clustered. Yes, this is O(n2)
- Form a new cluster with the nearby points.
  - I push a new Cluster to the entries array.
- Choose a new point that isn’t part of a cluster, and repeat until we have visited all the points.
  - I choose the next non-processed marker.

With respect to performance, isn't the existing overlap-based clustering also O(n2) in the worst case (when none overlap)?

I modified DTA to use thumbnails that are about the size a user might want to display. Please see the attached videos that show the two algorithms. In the overlap clustering video we can see how marker "disappears" when it's clustered with its neighbor. 

https://user-images.githubusercontent.com/48103957/145650752-b5f233af-0a10-47b2-9a59-389b31653a47.mov

https://user-images.githubusercontent.com/48103957/145650719-7c887866-038c-4174-86da-3266d4af28b1.mov

Turning on the FPS option, I don't see much different between the two algorithms.

Before implementing this algorithm I did try to modify the existing code but I couldn't get it to work. I just tried again (now that I have refactored the code) using the union of the marker rects for the cluster rect and that seems to work too.

Maybe we introduce two new algorithms? I can add this new class to the PR once I clean it up and determine a good name for it.

I see that Google Maps has a number of clustering algorithms (https://github.com/googlemaps/js-markerclusterer) and they show a demo that compares them (https://googlemaps.github.io/js-markerclusterer/public/algorithms/)

Reply:
Here's a video of the overlap algorithm modified as described above.

https://user-images.githubusercontent.com/48103957/145653985-5e7f8cc9-7c46-4670-9674-cea665d3596c.mov

Reply:
@toddsouthenbentley The "greedy" implementation mainly differs in that it uses the distance between locations of the markers, rather than overlapping test. That tends to cluster more aggressively, particularly in camera views where markers can be small. I think I like that approach better, but i don't think we need more than one algorithm. There's really no difference between the result of the clustering with the current implementation and what you've added (other than the new way is less efficient). 

Whether the cluster location takes the location of its "anchor point" or the middle of the cluster is arbitrary. Probably the middle is better.

I don't think we need more than on clustering "algorithm". But, there should be some way for the user to specify the radius for clustering.

Reply:
OK, I have updated the PR to just use the new clustering algorithm. MarkerSet now has a new protected property:

```ts
  /** The radius (in pixels) for clustering markers, default 0 which implies calculating the radius based on the visible marker imageSize/size. */
  protected clusterRadius = 0;
```
I also updated IncidentMarkerDemo in display-test-app to create the clustered marker at the middle:

```ts
class IncidentMarkerSet extends MarkerSet<IncidentMarker> {
  protected getClusterMarker(cluster: Cluster<IncidentMarker>): Marker {
    return new IncidentClusterMarker(cluster.getAverageLocation(), cluster.markers[0].size, cluster, IncidentMarkerDemo.decorator!.warningSign);
  }
}
```

There are 3 other implementations of `getClusterMarker` in the source tree, would it be better to update them too?

Reply:
@toddsouthenbentley I meant to keep the existing implementation, but use distance-between-positions. I pushed a change that does that.

Reply:
Ah, sorry about that, thanks for the commit. 

I fixed the distSquared initialization:
```ts
      let distSquared = this.clusterRadius * this.clusterRadius;
```

Reply:
@toddsouthenbentley BTW, thank you very much for this PR! I know it was a lot of work, but I think it's a good improvement that many people will appreciate.

Reply:
Always glad to help out!